vault-tool/src/Network/VaultTool/TOTP.hs

```diff
@@ -0,0 +1,252 @@
+    Authenticated,
+    VaultConnection,
+    VaultMountedPath (..),
+ )
+
+-- | The data needed for vault to generate a TOTP Key
+-- <https://www.vaultproject.io/api-docs/secret/totp#parameters>
+data GenerateKeyRequest = GenerateKeyRequest
+    { -- | The name of the key to create
+      gkrKeyName :: Text,
+      -- | The name of the key's issuing organization
+      gkrIssuer :: Text,
+      -- | The name of the account associated with the key
+      gkrAccountName :: Text,
+      -- | The size of the key in bytes (defaults: 20)
+      gkrKeySize :: Maybe Int,
+      -- | The length of time in seconds used to generate a counter for the TOTP code calculation (default: 30)
+      gkrPeriod :: Maybe Int,
+      -- | The hashing algorithm to use. Options are 'SHA1', 'SHA256', 'SHA512'. (default: 'SHA1')
+      gkrAlgorithm :: Maybe HashAlgorithm,
+      -- | The number of digits in the generated TOTP code. Options are 'SixDigits' or 'EightDigits'. (default: 'SixDigits')
+      gkrDigitCount :: Maybe DigitCount,
+      -- | The number of delay periods allowed when validating TOTP code. Options are 'NoSkew' or 'OnePeriodSkew'.
+      -- (default: 'OnePeriodSkew')
+      gkrSkew :: Maybe Skew,
+      -- | The pixel size of the square QR code when generating a new key (default: 200)
+      gkrQrSize :: Maybe Int
+    }
+    deriving (Show, Eq)
+
+instance ToJSON GenerateKeyRequest where
+    toJSON x = object
+        [ "generate" .=! True
+        , "issuer" .=! gkrIssuer x
+        , "account_name" .=! gkrAccountName x
+        , "key_size" .=? gkrKeySize x
+        , "period" .=? gkrPeriod x
+        , "algorithm" .=? gkrAlgorithm x
+        , "digits" .=? gkrDigitCount x
+        , "skew" .=? gkrSkew x
+        , "qr_size" .=? gkrQrSize x
+        ]
+
+data HashAlgorithm = SHA1 | SHA256 | SHA512
+    deriving (Show, Eq)
+
+instance FromJSON HashAlgorithm where
+    parseJSON (String "SHA1") = pure SHA1
+    parseJSON (String "SHA256") = pure SHA256
+    parseJSON (String "SHA512") = pure SHA512
+    parseJSON x = fail $ "Expected 'SHA1', 'SHA256' or 'SHA512' but received " <> show x
+
+instance ToJSON HashAlgorithm where
+    toJSON = \case
+        SHA1 -> String "SHA1"
+        SHA256 -> String "SHA256"
+        SHA512 -> String "SHA512"
+
+data DigitCount = SixDigits | EightDigits
+    deriving (Show, Eq)
+
+instance FromJSON DigitCount where
+    parseJSON (Number 6) = pure SixDigits
+    parseJSON (Number 8) = pure EightDigits
+    parseJSON x = fail $ "Expected 6 or 8 but received " <> show x
+
+instance ToJSON DigitCount where
+    toJSON = \case
+        SixDigits -> Number 6
+        EightDigits -> Number 8
+
+data Skew = NoSkew | OnePeriodSkew
+    deriving (Show, Eq)
+
+instance ToJSON Skew where
+    toJSON = \case
+        NoSkew -> Number 0
+        OnePeriodSkew -> Number 1
+
+-- | The newly generated Vault TOTP key
+data GeneratedKey = GeneratedKey
+    { -- | The resulting base64 encoded QR code PNG
```

Comment:
If this is a data URI, rather than just the image data, it would be worth spelling that out.

Reply:
Is this what you're looking for? 9e02d47a59e64cd79523efe996696356ea255bef

Reply:
No, [this](https://en.wikipedia.org/wiki/Data_Uri)

Reply:
Oh, my bad. It's not a data URI, it's an `otpauth` URI. The link I have in the comments breaks down the structure of an otpauth URI.

Reply:
Let me clarify.  The first record field contains a "base64 encoded QR code".  Is this a data URI for a PNG image?